### PR TITLE
Lua 5.3 + 5.4 integer support, with CI and conflicts fixed

### LINKED ISSUE #36: Won't built against Lua >= 5.2
Will not build in Lua version >= 5.2 because lua_objlen() is used which is 5.1 only. It was [renamed to lua_rawlen() starting in 5.2](https://www.lua.org/manual/5.2/manual.html#8.3)  I can find no statement(s) within the Cjson documentation noting this nor providing/suggesting a workaround.

Build ...
```
lua_cjson.c:743:19: warning: implicit declaration of function 'lua_objlen' is invalid in C99 [-Wimplicit-function-declaration]
            len = lua_objlen(l, -1);
                  ^
1 warning generated.
```
Test ...
`$ lua -e "require 'cjson'"`

Results
```
/usr/local/bin/lua: error loading module 'cjson' from file '/usr/local/lib/lua/5.2/cjson.so':
	dlopen(/usr/local/lib/lua/5.2/cjson.so, 6): Symbol not found: _lua_objlen
  Referenced from: /usr/local/lib/lua/5.2/cjson.so
  Expected in: flat namespace
 in /usr/local/lib/lua/5.2/cjson.so
stack traceback:
	[C]: in ?
	[C]: in function 'require'
	(command line):1: in main chunk
	[C]: in ?
```

### LINKED ISSUE #67: LuaRocks version fails on Lua 5.4
Hello,

The rockspec published on LuaRocks installs the tag 2.1.0.6 which fails at execution on Lua 5.4:

```
lua_modules/lib/lua/5.4/cjson.so: undefined symbol: lua_objlen
```

There's also a warning at installation time about that:

```
Installing https://luarocks.org/lua-cjson-2.1.0.6-1.src.rock

lua-cjson 2.1.0.6-1 depends on lua >= 5.1 (5.4-1 provided by VM)
gcc -O2 -fPIC -I/usr/include -c lua_cjson.c -o lua_cjson.o
lua_cjson.c: In function 'json_append_data':
lua_cjson.c:743:19: warning: implicit declaration of function 'lua_objlen' [-Wimplicit-function-declaration]
             len = lua_objlen(l, -1);
                   ^
```

When installing from a locally-altered rockspec that removes the tag constraint (i.e. installing HEAD) works without issue. The fix appears to be in commit f5cf829 which was done after both 2.1.0.7 and 2.1.0.8, so I guess ideally a new version should be tagged and published.

Thanks,
Martin

### PR DESCRIPTION
**Pull request by hishamhm**
Hello @leafo, @agentzh and @cloudwu!

I just took #3 which has been open since 2015(!) and got it to a mergeable state, by reconciling the changes made in https://github.com/openresty/lua-cjson and https://github.com/cloudwu/lua-cjson. I also added Lua 5.4 to the test matrix added by @leafo in #50.

I have made an explicit merge commit to keep history of both forks, but I can also rebase the cloudwu branch onto this repo if preferred. I'd just like this to get merged and the lua-cjson Lua version compatibility nightmare to be over after so many years. :)

I also noticed that version 2.1.0.10 was tagged in this repo but it was never uploaded to LuaRocks: https://luarocks.org/modules/openresty/lua-cjson

Ideally, we should merge this, tag 2.1.0.11 and upload it to LuaRocks (I can either add a commit with the version bump, or make a separate PR).

Thanks!

---
Closes #3.
Fixes #67.
Fixes #36.

### CODE REVIEW

**Comment by hishamhm** [comment]
Also, friendly ping to @xiaocang since you tagged 2.1.0.10 :)

**Comment by leafo** [comment]
Just a heads up: I pushed 2.1.0.10 and 2.1.0.9 to luarocks.org today, which should allow it to install on Lua 5.2+.

**Comment by hishamhm** [comment]
@leafo I rebased the original version of this PR into latest `master`, and I got the Github Actions workflow to run green (at least on my fork; it does not show up here).

So this branch now contains a bunch of unrelated / semi-related things:

* Lua 5.3+ integer support, from @cloudwu + tweaks and fixes, including those of mine
* adjustments to get the Github Actions CI runnable and green (it's not activated in this repo though? it seems to have both Travis and Actions files, but only Travis is running here)
* vendored dtoa.c version bump from upstream (which I did while fixing cppcheck warnings from CI)
* various other minor upstream commits, from original lua-cjson author @mpx as well as from @cloudwu's fork
* rockspec fix and bump, readying it for a new release 2.1.0.14.

I can split these into multiple PRs if necessary for review but I think the Actions CI will only run green with all of them.

@Tieske has pinged me asking whether we can get a new release pushed to luarocks.org because of the security fix in 2.1.0.13 (which has the security fix but a broken rockspec). The latest version in luarocks.org is 2.1.0.10.

Could we get this PR reviewed/merged and a new release out? Or at the very least push 2.1.0.13 to luarocks.org with a fixed rockspec.

**Review by zhuizhuhaomeng — APPROVED** [review]

File: dtoa.c [hunk 1]
```diff
@@ -1533,12 +1533,18 @@ ThInfo {
...
 {
 	size_t L;
+	ThInfo *newTI1;
 
 	if (n > maxthreads) {
 		L = n*sizeof(ThInfo);
 		if (TI1) {
-			TI1 = (ThInfo*)REALLOC(TI1, L);
-			memset(TI1 + maxthreads, 0, (n-maxthreads)*sizeof(ThInfo));
+			newTI1 = (ThInfo*)REALLOC(TI1, L);
```
Comment (zhuizhuhaomeng) [inline dtoa.c:1541]: Would you please tell me the url of this dtoa.c?

**Review by zhuizhuhaomeng — CHANGES_REQUESTED** [review]

File: dtoa.c (hunk 1 above)
Comment (hishamhm) [reply dtoa.c:1541]: This is based on
https://portal.ampl.com/~dmg/netlib/fp/dtoa.c
https://portal.ampl.com/~dmg/netlib/fp/changes

For this particular change you highlighted, I wrote it myself (see commit [026466c](https://github.com/openresty/lua-cjson/pull/84/commits/026466cd0f51db95508b4cf353d3f9d936e529dd)), as it was an issue reported in CI by cppcheck.

**Pull request merged**